File: cli/src/global-config.ts
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

export interface GlobalConfig {
  registry?: string;
}

const DEFAULT_REGISTRY = 'ghcr.io/joacohbc/';

export function globalConfigDir(): string {
  if (process.platform === 'win32') {
    const appData = process.env.APPDATA;
    if (appData) return path.join(appData, 'devcontainer-cli');
    return path.join(os.homedir(), 'AppData', 'Roaming', 'devcontainer-cli');
  }
  const xdg = process.env.XDG_CONFIG_HOME;
  if (xdg) return path.join(xdg, 'devcontainer-cli');
  return path.join(os.homedir(), '.config', 'devcontainer-cli');
}

export function globalConfigPath(): string {
  return path.join(globalConfigDir(), 'config.json');
}

export function loadGlobalConfig(): GlobalConfig {
  const p = globalConfigPath();
  if (!fs.existsSync(p)) return {};
  try {
    const raw = fs.readFileSync(p, 'utf8');
    const parsed = JSON.parse(raw) as GlobalConfig;
    return parsed && typeof parsed === 'object' ? parsed : {};
  } catch {
    return {};
  }
}

export function saveGlobalConfig(cfg: GlobalConfig): void {
  const dir = globalConfigDir();
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(globalConfigPath(), JSON.stringify(cfg, null, 2) + '\n');
}

function normalizeRegistry(url: string): string {
  const trimmed = url.trim();
  if (!trimmed) return trimmed;
  return trimmed.endsWith('/') ? trimmed : `${trimmed}/`;
}

export function resolveRegistry(flagOverride?: string, perProject?: string): string {
  if (flagOverride) return normalizeRegistry(flagOverride);
  if (perProject) return normalizeRegistry(perProject);
  const global = loadGlobalConfig().registry;
  if (global) return normalizeRegistry(global);
  return DEFAULT_REGISTRY;
}

export const REGISTRY_DEFAULT = DEFAULT_REGISTRY;

File: cli-go/internal/domain/global_config.go
package domain

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

type GlobalConfig struct {
	Registry string `json:"registry,omitempty"`
}

const DefaultRegistry = "ghcr.io/joacohbc/"

func GlobalConfigDir() string {
	if runtime.GOOS == "windows" {
		appdata := os.Getenv("APPDATA")
		if appdata != "" {
			return filepath.Join(appdata, "devcontainer-cli")
		}
		return filepath.Join(os.Getenv("USERPROFILE"), "AppData", "Roaming", "devcontainer-cli")
	}
	// Read XDG_CONFIG_HOME at call time so tests can override it with os.Setenv.
	xdgConfigHome := os.Getenv("XDG_CONFIG_HOME")
	if xdgConfigHome == "" {
		home, _ := os.UserHomeDir()
		xdgConfigHome = filepath.Join(home, ".config")
	}
	return filepath.Join(xdgConfigHome, "devcontainer-cli")
}

func GlobalConfigPath() string {
	return filepath.Join(GlobalConfigDir(), "config.json")
}

func LoadGlobalConfig() GlobalConfig {
	data, err := os.ReadFile(GlobalConfigPath())
	if err != nil {
		return GlobalConfig{}
	}
	var cfg GlobalConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		return GlobalConfig{}
	}
	return cfg
}

func SaveGlobalConfig(cfg GlobalConfig) error {
	dir := GlobalConfigDir()
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	return os.WriteFile(GlobalConfigPath(), data, 0644)
}

func normalizeRegistry(url string) string {
	trimmed := strings.TrimSpace(url)
	if trimmed == "" {
		return trimmed
	}
	if !strings.HasSuffix(trimmed, "/") {
		return trimmed + "/"
	}
	return trimmed
}

func ResolveRegistry(flagOverride, perProject string) string {
	if flagOverride != "" {
		return normalizeRegistry(flagOverride)
	}
	if perProject != "" {
		return normalizeRegistry(perProject)
	}
	cfg := LoadGlobalConfig()
	if cfg.Registry != "" {
		return normalizeRegistry(cfg.Registry)
	}
	return DefaultRegistry
}
